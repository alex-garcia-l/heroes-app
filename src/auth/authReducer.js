import { authType } from '../types/auth';

export const authReducer = (state = {}, action) => {
  
  switch (action.type) {
    case authType.login:
      return {
        ...action.payload,
        logged: true
      }
    case authType.logout:
      return {
        logged: false
      }
  
    default:
      return state;
  }

}