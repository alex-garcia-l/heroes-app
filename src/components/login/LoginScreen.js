import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';

import { AuthContext } from '../../auth/authContext';
import { authType } from '../../types/auth';
import './login.css';

export const LoginScreen = () => {

  const navigate = useNavigate();
  const date = new Date();
  const { userDispatch } = useContext(AuthContext);

  const handleSubmit = (evt) => {
    evt.preventDefault();

    const action = {
      type: authType.login,
      payload: {
        name: 'My name'
      }
    }

    userDispatch(action);
    const lastPath = localStorage.getItem('last_path') || '/home';

    navigate(lastPath, {
      replace: true
    });
  }

  return (
    <div className="text-center d-flex align-items-center vh-100">
      <main className="form-signin">
        <form onSubmit={handleSubmit}>
          <img className="mb-4" src="https://getbootstrap.com/docs/5.0/assets/brand/bootstrap-logo.svg" alt="" width="72" height="57" />
          <h1 className="h3 mb-3 fw-normal">Please sign in</h1>

          {/* <div className="form-floating">
            <input type="text" className="form-control" placeholder="Name" />
            <label htmlFor="floatingInput">Name</label>
          </div>

          <div className="form-floating">
            <input type="password" className="form-control" placeholder="Password" />
            <label htmlFor="floatingPassword">Password</label>
          </div> */}

          <button
            className="w-100 btn btn-lg btn-primary"
            type="submit"
          >
            Sign in
          </button>

          <p className="mt-5 mb-3 text-muted">Â© {date.getFullYear()}</p>
        </form>
      </main>
    </div>
  );
};
