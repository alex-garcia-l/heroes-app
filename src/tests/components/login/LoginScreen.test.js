import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { AuthContext } from '../../../auth/authContext';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { authType } from '../../../types/auth';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}));

describe('Prueba para el componente <LoginScreen />', () => {

  const action = {
    type: authType.login,
    payload: {
      name: 'My name'
    }
  }

  const contextValue = {
    user: {
      logged: false
    },
    userDispatch: jest.fn()
  }

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={['/login']}>
        <Routes>
          <Route path="login" element={<LoginScreen />} />
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe realizar el userDispatch y la navegaciÃ³n', () => {

    const handleSubmit = wrapper.find('form').prop('onSubmit');
    
    handleSubmit({
      preventDefault() { }
    });

    expect(contextValue.userDispatch).toHaveBeenCalledWith(action);
    expect(mockNavigate).toHaveBeenCalledWith('/home', {
      replace: true
    });

    const lastPath = '/marvel';
    localStorage.setItem('last_path', lastPath);

    handleSubmit({
      preventDefault() { }
    });

    expect(mockNavigate).toHaveBeenCalledWith(lastPath, {
      replace: true
    });

  });

});
