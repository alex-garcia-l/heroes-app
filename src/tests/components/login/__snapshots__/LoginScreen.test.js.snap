// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba para el componente <LoginScreen /> debe de mostrarse correctamente 1`] = `
<div
  className="text-center d-flex align-items-center vh-100"
>
  <main
    className="form-signin"
  >
    <form
      onSubmit={[Function]}
    >
      <img
        alt=""
        className="mb-4"
        height="57"
        src="https://getbootstrap.com/docs/5.0/assets/brand/bootstrap-logo.svg"
        width="72"
      />
      <h1
        className="h3 mb-3 fw-normal"
      >
        Please sign in
      </h1>
      <button
        className="w-100 btn btn-lg btn-primary"
        type="submit"
      >
        Sign in
      </button>
      <p
        className="mt-5 mb-3 text-muted"
      >
        Â© 
        2022
      </p>
    </form>
  </main>
</div>
`;
